librz/core/cmd_debug.c

```diff
@@ -1443,376 +1377,465 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+	} else if (argc == 2) { // dmp <perms>
+		addr = UT64_MAX;
 		rz_list_foreach (core->dbg->maps, iter, map) {
-			if (addr >= map->addr && addr < map->addr_end) {
-				rz_debug_map_dealloc(core->dbg, map);
-				rz_debug_map_sync(core->dbg);
-				return true;
+			if (core->offset >= map->addr && core->offset < map->addr_end) {
+				addr = map->addr;
+				size = map->size;
+				break;
 			}
 		}
-		eprintf("The address doesn't match with any map.\n");
-		break;
-	case 'L': // "dmL"
-	{
-		int size;
-		char *p = strchr(input + 2, ' ');
-		if (p) {
-			*p++ = 0;
-			addr = rz_num_math(core->num, input + 1);
-			size = rz_num_math(core->num, p);
-			rz_debug_map_alloc(core->dbg, addr, size, true);
+		perms = rz_str_rwx(argv[1]);
+		if (addr != UT64_MAX && perms >= 0) {
+			rz_debug_map_protect(core->dbg, addr, (int)size, perms);
 		} else {
-			eprintf("Usage: dmL addr size\n");
-			return false;
+			return RZ_CMD_STATUS_WRONG_ARGS;
 		}
-	} break;
-	case '\0': // "dm"
-	case '*': // "dm*"
-	case 'j': // "dmj"
-	case 'q': // "dmq"
-		rz_debug_map_sync(core->dbg); // update process memory maps
-		rz_debug_map_list(core->dbg, core->offset, input);
-		break;
-	case '=': // "dm="
-		rz_debug_map_sync(core->dbg);
-		rz_debug_map_list_visual(core->dbg, core->offset, input,
-			rz_config_get_i(core->config, "scr.color"));
-		break;
-	case 'h': // "dmh"
-		(void)rz_debug_heap(core, input);
-		break;
+	} else {
+		return RZ_CMD_STATUS_WRONG_ARGS;
 	}
-	return true;
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_debug_dmS_handler(RzCore *core, int argc, const char **argv, RzOutputMode m) {
+	if (rz_debug_is_dead(core->dbg)) {
+		rz_cons_println("Debugging is not enabled. Run ood?");
+		return RZ_CMD_STATUS_ERROR;
+	}
+	RzListIter *iter;
+	RzDebugMap *map;
+	ut64 addr;
+	const char *libname = NULL, *sectname = NULL, *mode = "";
+	ut64 baddr = 0LL;
+	if (m == RZ_OUTPUT_MODE_RIZIN) {
+		mode = "-r ";
+	}
+	addr = UT64_MAX;
+	if (argc == 3) {
+		sectname = argv[2];
+	}
+	if (argc >= 2) {
+		if (IS_DIGIT(*argv[1])) {
+			const char *a0 = argv[1];
+			addr = rz_num_math(core->num, a0);
+		} else {
+			addr = UT64_MAX;
+		}
+		if (!addr || addr == UT64_MAX) {
+			libname = argv[1];
+		}
+	}
+	rz_debug_map_sync(core->dbg); // update process memory maps
+	RzList *list = rz_debug_modules_list(core->dbg);
+	rz_list_foreach (list, iter, map) {
+		if ((!libname ||
+			    (addr != UT64_MAX && (addr >= map->addr && addr < map->addr_end)) ||
+			    (libname != NULL && (strstr(map->name, libname))))) {
+			baddr = map->addr;
+			char *res;
+			const char *file = map->file ? map->file : map->name;
+			char *name = rz_str_escape((char *)rz_file_basename(file));
+			char *filesc = rz_str_escape(file);
+			/* TODO: do not spawn. use RzBin API */
+			if (sectname) {
+				char *sect = rz_str_escape(sectname);
+				res = rz_sys_cmd_strf("env RZ_BIN_PREFIX=\"%s\" rz-bin %s-B 0x%08" PFMT64x " -S \"%s\" | grep \"%s\"", name, mode, baddr, filesc, sect);
```

Comment:
@thestr4ng3r could you please suggest a proper API as a substitute for this command?

Reply:
this is not ok.

Reply:
but we can change this in another PR